array_analyzer/extract/metadata.py

```diff
@@ -0,0 +1,221 @@
+
+        elif c.METADATA_EXTENSION == 'well':
+            self._set_run_path(output_folder_)
+            return
+
+        elif c.METADATA_EXTENSION == 'csv':
+            # check that three .csvs exist
+            three_csvs = ['array_format_antigen', 'array_format_type', 'array_parameters']
+            csvs = [f for f in os.listdir(input_folder_) if '.csv' in f]
+            if len(csvs) != 3:
+                raise IOError("incorrect number of .csv files found, aborting")
+            for target in three_csvs:
+                if True not in [target in file for file in csvs]:
+                    raise IOError(f".csv file with substring {target} is missing")
+
+            csv_paths = [os.path.join(input_folder_, one_csv) for one_csv in csvs]
+
+            # parsing .csv
+            fiduc_, _, repl_, params_ = txt_parser.create_csv_dict(csv_paths)
+
+        elif c.METADATA_EXTENSION == 'xlsx':
+            # check that exactly one .xlsx exists
+            # xlsxs = [f for f in os.listdir(input_folder_) if '.xlsx' in f]
+            # if len(xlsxs) > 1:
+            #     raise IOError("more than one .xlsx file found, aborting")
+            # xlsx_path = os.path.join(input_folder_, xlsxs[0])
+
+            # check that properly named .xlsx exists
+            if not os.path.isfile(os.path.join(input_folder_, 'Metadata_and_Plate_configuration.xlsx')):
+                raise IOError("required metadata file named 'Metadata_and_Plate_configuration.xlsx' does not exist")
+            xlsx_path = os.path.join(input_folder_, 'Metadata_and_Plate_configuration.xlsx')
+
+            # check that the xlsx file contains necessary worksheets
+            sheets = pd.read_excel(xlsx_path, sheet_name=None)
+            if 'imaging_and_array_parameters' not in sheets.keys():
+                raise IOError("sheet by name 'imaging_and_array_parameters' not present in excel file, aborting")
+            if 'array_antigens' not in sheets.keys():
+                raise IOError("sheet by name 'array_antigens' not present in excel file, aborting")
+
+            # parsing .xlsx
+            fiduc_, _, repl_, params_ = txt_parser.create_xlsx_dict(xlsx_path)
+
+        else:
+            raise NotImplementedError(f"metadata with extension {c.METADATA_EXTENSION} is not supported")
+
+        # setting constants
+        if fiduc_:
+            c.fiducials = fiduc_
+        if spots_:
+            c.spots = spots_
+        if repl_:
+            c.replicates = repl_
+        if params_:
+            c.params['rows'] = int(params_['rows'])
+            c.params['columns'] = int(params_['columns'])
+            c.params['v_pitch'] = float(params_['v_pitch'])
+            c.params['h_pitch'] = float(params_['h_pitch'])
+            c.params['spot_width'] = float(params_['spot_width'])
+            if c.METADATA_EXTENSION == 'xml':
+                c.params['pixel_size'] = c.params['pixel_size_scienion']
```

Comment:
Just for my understanding: is this because scienion is the only acquisition with xml metadata so that pixel size is known?

Reply:
yes that is exactly right.  As far as I know, the sciReader doesn't even have other magnifications.  We've always seen the same pixel size